File: jane-kim-279/week_5/271p.py
from itertools import combinations
from collections import Counter

def solution(orders, course):
    answer = []
    
    for length in course:
        candidates = []
        for order in orders:
            order = sorted(order) # 주문을 정렬하여 'AB'와 'BA'를 같은 것으로 처리
            combi = combinations(order, length)
            candidates.extend(''.join(c) for c in combi)
        
        # 각 조합이 몇 번 등장했는지 세기
        menu_count = {}
        for candidate in candidates:
            if candidate in menu_count:
                menu_count[candidate] += 1
            else:
                menu_count[candidate] = 1
        
        # menu_count가 비어있지 않고, 가장 많이 등장한 횟수가 2 이상인 경우
        if menu_count and max(menu_count.values()) >= 2:
            max_count = max(menu_count.values())
            for menu, count in menu_count.items():
                if count == max_count:
                    answer.extend([menu])
    
    return sorted(answer)

File: mint723/week_5/271p.java
// https://school.programmers.co.kr/learn/courses/30/lessons/72411

import java.util.*;

class Solution {
    static Map<Integer, Map<String,Integer>> courseMap;
    
    public List<String> solution(String[] orders, int[] course) {
        courseMap = new HashMap<>();
        for(int i : course){
            courseMap.put(i, new HashMap<>());
        }
        
        // 각 주문에 대해 모든 가능한 조합 생성
        for(String order : orders){
            char[] orderArray = order.toCharArray();
            // 조합에서 AB, BA는 같은 코스로 체크해야함
            Arrays.sort(orderArray);
            combinations(0, orderArray, "");
        }
        
        List<String> answer = new ArrayList<>();
        
        // 각 코스 길이별로 정리
        for(Map<String, Integer> count : courseMap.values()){
            count.values()
                .stream()
                .max(Comparator.comparingInt(i -> i)) // 가장 빈도수가 높은 코스
                .ifPresent(cnt -> count.entrySet() // 코스에 대한 메뉴 수가 가능할 때만
                         .stream()
                         .filter(entry -> cnt.equals(entry.getValue()) && cnt>1) // 2명 이상의 손님으로부터 주문된 단품 메뉴 조합
                         .forEach(entry -> answer.add(entry.getKey()))); 
        }
        
        Collections.sort(answer);
        
        return answer;
    }
    
    // 조합 메서드
    static void combinations(int idx, char[] order, String result){
        if(courseMap.containsKey(result.length())){
            Map<String, Integer> map = courseMap.get(result.length());
            
            map.put(result, map.getOrDefault(result, 0) + 1);
        }
        
        for(int i=idx; i<order.length; i++){
            combinations(i + 1, order, result + order[i]);
        }
    }
}
